hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Acinetobacter\_baumannii -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Acinetobacter\_baumannii.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Acinetobacter\_pittii -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Acinetobacter\_pittii.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Acinetobacter\_radioresistens -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Acinetobacter\_radioresistens.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Citrobacter\_portucalensis -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Citrobacter\_portucalensis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Enterobacter\_cancerogenus -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Enterobacter\_cancerogenus.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Enterobacter\_sp.\_SA187 -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Enterobacter\_sp.\_SA187.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Leuconostoc\_citreum -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Leuconostoc\_citreum.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Leuconostoc\_pseudomesenteroides -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Leuconostoc\_pseudomesenteroides.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Pseudomonas\_oryzihabitans -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Pseudomonas\_oryzihabitans.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Serratia\_symbiotica -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Serratia\_symbiotica.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Sodalis\_glossinidius -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Sodalis\_glossinidius.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Sodalis\_phage\_phiSG1 -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Sodalis\_phage\_phiSG1.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Sodalis\_praecaptivus -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Sodalis\_praecaptivus.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Staphylococcus\_capitis -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Staphylococcus\_capitis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Streptococcus\_gordonii -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Streptococcus\_gordonii.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Arsenophonus\_nasoniae -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Arsenophonus\_nasoniae.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Bartonella\_apis -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Bartonella\_apis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Cutibacterium\_acnes -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Cutibacterium\_acnes.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Frischella\_perrara -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Frischella\_perrara.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Gilliamella\_apicola -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Gilliamella\_apicola.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Lactobacillus\_kunkeei -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Lactobacillus\_kunkeei.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Lactococcus\_lactis -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Lactococcus\_lactis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Pantoea\_agglomerans -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Pantoea\_agglomerans.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Pantoea\_ananatis -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Pantoea\_ananatis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Tatumella\_sp.\_TA1 -U  ~/elixir/metagenomics/SPlab/reads/R_2020_02_12_12_23_36_user_IONAS-440-PMN-SPlab-DrMylonis_200212_PMNcAmp2-6_SPD5-10_IM3R7-10.SPD10_DSM-H7.IonXpress_010.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD10_Tatumella\_sp.\_TA1.bam
