hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Acinetobacter\_baumannii -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Acinetobacter\_baumannii.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Acinetobacter\_pittii -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Acinetobacter\_pittii.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Acinetobacter\_radioresistens -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Acinetobacter\_radioresistens.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Citrobacter\_portucalensis -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Citrobacter\_portucalensis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Enterobacter\_cancerogenus -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Enterobacter\_cancerogenus.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Enterobacter\_sp.\_SA187 -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Enterobacter\_sp.\_SA187.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Leuconostoc\_citreum -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Leuconostoc\_citreum.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Leuconostoc\_pseudomesenteroides -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Leuconostoc\_pseudomesenteroides.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Pseudomonas\_oryzihabitans -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Pseudomonas\_oryzihabitans.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Serratia\_symbiotica -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Serratia\_symbiotica.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Sodalis\_glossinidius -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Sodalis\_glossinidius.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Sodalis\_phage\_phiSG1 -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Sodalis\_phage\_phiSG1.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Sodalis\_praecaptivus -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Sodalis\_praecaptivus.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Staphylococcus\_capitis -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Staphylococcus\_capitis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Streptococcus\_gordonii -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Streptococcus\_gordonii.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Arsenophonus\_nasoniae -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Arsenophonus\_nasoniae.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Bartonella\_apis -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Bartonella\_apis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Cutibacterium\_acnes -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Cutibacterium\_acnes.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Frischella\_perrara -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Frischella\_perrara.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Gilliamella\_apicola -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Gilliamella\_apicola.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Lactobacillus\_kunkeei -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Lactobacillus\_kunkeei.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Lactococcus\_lactis -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Lactococcus\_lactis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Pantoea\_agglomerans -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Pantoea\_agglomerans.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Pantoea\_ananatis -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Pantoea\_ananatis.bam
hisat2 -k 1 -p2 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8 -x ~/Elixir/metagenomics/species/idx/Tatumella\_sp.\_TA1 -U  ~/elixir/metagenomics/SPlab/reads/R_2019_08_07_16_07_46_user_IONAS-431-GKlab-MS_SPlab_190807_GK3R288-302_SP3R1-4.SPD3_fragDNA1.IonXpress_031.fastq |  /data/results/tools/samtools/samtools-1.3/samtools view -b -F 4 | /data/results/tools/samtools/samtools-1.3/samtools  sort -@ 2 -o SPD3_Tatumella\_sp.\_TA1.bam
