---
title: "VARROA CORRELATION WITH METAGENOMIC AND ALIGNED SEQUENCES "
  chunk_output_type: inline
---

I create a excel file with:
- the october 2020 results of metagenomic quantification performed by Anastasio after kraken filtering and normalisation using DESeq2 taking into account the method and season. 
- the reads aligned against varroa destructor genome performed by Martin Rezlo using stricter mapping parameters the settings for the hisat2 mapper to avoid any mismatches
(-k 1 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8)
- the average and sum of varroa count from the apiary monitoring (Varday) from the natural fall of varroa (see APIARY_IND.R for detailed of the calculation. average over 2weeks, 1 month, 2 months, 2 months. 

To get an overview of the situation I made a general correlation matrix for each varroa quantification
CPM= count per million (normalised aligned reads against genome)


```{r}
setwd("~/Documents/GitHub/Direct-shotgun-metagenomics-pub/Varroa 2")

library(readxl) # import xlsx file
library(ggplot2)
library(ggcorrplot)

#Import data
# Varroa_genus is the data at genus level
varroa2<-read_excel("Varroa_genus.xlsx")

# clean matrix
varroa2_mat <- dplyr::select_if(varroa2, is.numeric)

# log2 transform CPMs
# I was reading that usually people log2 transform cpm values this way.
# If needed you can apply the same function for Deseq data, I have only log2 transformed them.
log_reads <- function(x) {
  log2(x+1)
}

# calculate log2 values
varroa2_mat$CPM_DSM <- log_reads(varroa2_mat$CPM_DSM)
varroa2_mat$CPM_SM <- log_reads(varroa2_mat$CPM_SM)
varroa2_mat$DeSeq2_DSM <- log2(varroa2_mat$DeSeq2_DSM)
varroa2_mat$DeSeq2_SM <- log2(varroa2_mat$DeSeq2_SM)


# calculate the correlations
r <- cor(varroa2_mat)
round(r,2)



#correlation matrix
ggcorrplot(r, 
           hc.order = TRUE, 
           type = "lower",
           lab = TRUE)


```

Results: The metagenomic data SM correlate the most with 2 weeks varroa monitoring, then 1 month then 2months but impressively correlate inversely with the CMP. Something that doesn't make sense at all...
The metagenomic data Direct SM correlate the most with the 3 months varroa monitoring, and impressively correlate inversely with the 2 weeks and 1 month and 2 month and not al all with the CMP. Nothing make sense. 
How come that SM Direct SM seems to anti-correlate if they were not significant in Anastasio analysis? For the DeSeq2 normalized, the counts were very little, is it because of the kraken filter? 
I take the aligned read from the log file of Martin, is it the correct one? Need to verify with bam files



