---
title: "VARROA CORRELATION WITH METAGENOMIC AND ALIGNED SEQUENCES "
  chunk_output_type: inline
---

I create a excel file with:
- the october 2020 results of metagenomic quantification performed by Anastasio after kraken filtering and normalisation using DESeq2
- the reads aligned against varroa destructor genome performed by Martin Rezlo using stricter mapping parameters the settings for the hisat2 mapper to avoid any mismatches
(-k 1 --dta --no-spliced-alignment --rdg 10,6 --rfg 10,6 --mp 24,8)
- the average and sum of varroa count from the apiary monitoring (Varday) from the natural fall of varroa (see APIARY_IND.R for detailed of the calculation. average over 2weeks, 1 month, 2 months, 3 months. 

To get an overview of the situation I made a general correlation matrix for each varroa quantification
CPM= count per million (normalised aligned reads against genome)


```{r}
setwd("~/Documents/GitHub/Direct-shotgun-metagenomics-pub/Varroa 2")

library(readxl) # import xlsx file
library(ggplot2)
library(ggcorrplot)
library(data.table)
library(reshape2)
library(viridis)
library(dplyr)

#Import data
# Varroa_genus is the metagenomic data at genus level
varroa2<-read_excel("Varroa_genus.xlsx")
varroa2<-varroa2[,-c(2:3,6:7,13)]

# clean matrix
varroa2_mat <- dplyr::select_if(varroa2, is.numeric)


# log2 transform CPMs
# I was reading that usually people log2 transform cpm values this way.
# If needed you can apply the same function for Deseq data, I have only log2 transformed them.
log_reads <- function(x) {
  log2(x+1)
}

varroa2_mat<-log_reads(varroa2_mat)

#### OVERALL
# calculate the correlations
r <- cor(varroa2_mat)
round(r,2)

#correlation matrix
ggcorrplot(r, 
           hc.order = TRUE, 
           type = "lower",
           lab = TRUE)

#####SM
varroa2_mat_SM<-varroa2_mat[,-c(1:2)]
r_SM <- cor(varroa2_mat_SM)
round(r_SM,2)
ggcorrplot(r_SM, 
           hc.order = TRUE, 
           type = "lower",
           lab = TRUE)
####DSM
varroa2_mat_DSM<-varroa2_mat[,-c(3:4)]
r_DSM <- cor(varroa2_mat_DSM)
round(r_DSM,2)
ggcorrplot(r_DSM, 
           hc.order = TRUE, 
           type = "lower",
           lab = TRUE)
```

Results: The metagenomic data SM correlate the most with 2 weeks varroa monitoring, then 1 month then 2months but impressively correlate inversely with the CMP. Something that doesn't make sense at all...
The metagenomic data Direct SM correlate the most with the 3 months varroa monitoring, and impressively correlate inversely with the 2 weeks and 1 month and 2 month and not al all with the CMP. Nothing make sense. 
How come that SM Direct SM seems to anti-correlate if they were not significant in Anastasio analysis? For the DeSeq2 normalized, the counts were very little, is it because of the kraken filter? 


Final plots
```{r}
##### FINAL PLOTS 
#import organised  data 
Varroa_final<-read_excel("Varroa_genus.xlsx", sheet="Final", col_types = c("text","text","numeric", "text","numeric"))

Varroa_final<- Varroa_final%>%
  filter(!Monitoring=="average 3 months")

#transform read count in log
Varroa_final$Reads<-log_reads(Varroa_final$Reads)
Varroa_final$Natural_Fall<-log_reads(Varroa_final$Natural_Fall)

#plot
ggscatter( Varroa_final, x = "Reads", y = "Natural_Fall", color = "Monitoring", palette = "jco", add = "reg.line", shape="Hive") +
  scale_color_viridis(discrete=TRUE) +
  theme_classic() +
  facet_wrap(~ Method,scales = "free") +
  theme(plot.title = element_text(size=14), 
        axis.text=element_text(size=12,colour = "black"),
        axis.title=element_text(size=12),
        legend.title = element_blank(),
        legend.text = element_text(size = 12))+
  ggtitle("Metagenomic correlation with Natural Varroa fall" ) +
  ylab("Varroa count/day (log)") +
  xlab("Normalised reads count (log)") +
  stat_cor(aes(color = Monitoring))

```


